### PR TITLE
WorkQueue: Use rlock for reporting to allow concurrent readers

### PR DESCRIPTION
**Pull request by gaurav1086**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Use read lock to allow multiple goroutines to report concurrently.
A RWMutex is a reader/writer mutual exclusion lock. The lock can be held by an arbitrary number of readers or a single writer. The zero value for a RWMutex is an unlocked mutex. In other words, readers don't have to wait for each other. They only have to wait for writers holding the lock.

A sync.RWMutex is thus preferable for data that is mostly read, and the resource that is saves compared to a sync.Mutex is time.
Reference: https://golang.org/pkg/sync/#RWMutex.Lock

**Which issue(s) this PR fixes**:

Fixes #
N/A
**Special notes for your reviewer**:
N/A
**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by fedebongio** [comment]
Hi Gaurav, thanks for your PR. Can you also provide metrics about the improvements that this PR achieves before reviewing it? 
/cc @lavalamp @deads2k

**Comment by gaurav1086** [comment]
> Hi Gaurav, thanks for your PR. Can you also provide metrics about the improvements that this PR achieves before reviewing it?
> /cc @lavalamp @deads2k

@fedebongio thanks. I updated the PR.

**Comment by lavalamp** [comment]
Is there actually contention over this lock? Did you find this by profiling the code? We're fairly reluctant as a rule to make locking changes that don't have benchmark or profile demonstrating contention or other benefit.

**Comment by gaurav1086** [comment]
> Is there actually contention over this lock? Did you find this by profiling the code? We're fairly reluctant as a rule to make locking changes that don't have benchmark or profile demonstrating contention or other benefit.

@lavalamp , thanks for the comment. 

"Is there actually contention over this lock? " Yes, because there are already mutex locks in place which indicates that its a critical section for multiple goroutines. The change is not introducing any **new locks** but is essentially only changing the mutex type to RWMutex and **replacing full vanilla locks with reader's locks for concurrent reading**. This is an optimization to avoid contention amongst multiple readers to wait for each other.

**Comment by lavalamp** [comment]
I understand what this change is doing.

We do not modify locks without a demonstration that there is actually contention in real-world conditions. E.g., you need to take a contention profile from `debug/pprof/block` in a running controller manager and demonstrate that a non-negligible amount of time is spent blocked on acquiring this lock, and that this change reduces it meaningfully. Alternatively, you could write a benchmark that demonstrates the performance improvement.

(My expectation, to be clear, is that the method is almost never called and therefore this change will not in practice reduce contention. But I would be happy to be wrong.)

We have this high bar because locks are more or less the trickiest part of the codebase and it's very easy to do something that looks safe but isn't.

**Comment by gaurav1086** [comment]
@lavalamp thanks again for the explanation. Unfortunately, currently I do not have a setup for profiling this. However below are my two cents.

_"We do not modify locks without a demonstration that there is actually contention in real-world conditions."_ - This statement will apply here if I was doing the other way around i.e. changing a sync.RWMutex to sync.Mutex . 

_"We have this high bar because locks are more or less the trickiest part of the codebase and it's very easy to do something that looks safe but isn't."_ - Totally get it however it clearly looks from the code that its safe as we are only changing the nature/granularity of the lock while reading as documented in the Go manual.

Lastly, from the coding standard point of view, its a pretty standard practice to always have a reader's lock when only reading from the shared data and a full (writer's) lock when writing to it.

**Comment by lavalamp** [comment]
> "We do not modify locks without a demonstration that there is actually contention in real-world conditions." - This statement will apply here if I was doing the other way around i.e. changing a sync.RWMutex to sync.Mutex .

Correct. Any change to locking is going to get this response from me.

> "We have this high bar because locks are more or less the trickiest part of the codebase and it's very easy to do something that looks safe but isn't." - Totally get it however it clearly looks from the code...

Changes in this category are generally easy to produce and difficult to review, with many of them containing subtle bugs. Requiring justification like I asked for levels the playing field and makes the review and writing costs more equal. Basically, we have had a number of changes like this one in the past, and they have added up to an unintentional DoS attack on reviewers. I agree that yours looks safe at first glance, but I haven't looked carefully, and given the category it matches, I can't approve it without a strong objective reason because it would be inconsistent with past requests we've made of other contributors.

Basically, IMO, the bar should be the same for all changes in this category. "Looks safe at first glance and theoretically better" isn't good enough, we need "is objectively and non-neglibibly better as shown by profile comparison or representative benchmark, etc".

We can ask @deads2k for a second opinion.

**Comment by deads2k** [comment]
I understand the desire to have the prettiest or best theoretical locking code, but unless we see contention in some reasonable profile I'm disinclined to make changes to how concurrency is managed in our code. This is really the case anywhere in our codebase, but especially in error paths (which this is).

If you can demonstrate a bug or if you have a runtime profile that indicates you're getting significant contention here, I'm happy to reconsider.

/hold

**Comment by gaurav1086** [comment]
@deads2k @lavalamp thanks for the valuable pointers. Ran a benchmark with sync.Mutex vs sync.RWMutex and on the contrary found some conflicting results. It appears that the mutex.Lock() is 2x faster than the RWMutex.RLock() 

Description: Two scenarios (with lock and rwlock) with 1000 goroutines contending to read data from a slice (size = 10) through randomly generated index and write the value to separate files.

It turns out that  sync.Mutex::Lock() is faster than sync.RWMutex::RLock() by~ 2X.

Locks			RWLocks
2.589610687s		5.52346426s
2.693844854s		6.13264221s
2.574230731s		6.857576762s

Code snippet: https://play.golang.com/p/RZ-bE3OJ4H5 . 
Looking forward to your thoughts on this.

Go issue: https://github.com/golang/go/issues/38813

**Comment by gaurav1086** [comment]
Sorry for the late response. So I did some more research on this and while ItemFastSlowRateLimiter.NumRequeuest() is not called as often, however ItemExponentialFailureRateLimiter.NumRequeues() is called quite a lot of times and can possibly have enough contentions though I don't have a direct way of simulating/proving this.

**ItemFastSlowRateLimiter.NumRequeues()** is only called in a test file: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters_test.go

**ItemExponentialFailureRateLimiter.NumRequeuest()** : 
1. function: UpdateAPIService in file: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go and subequently UpdateAPIService()

2. function syncJob() and getBackoff() pkg/controller/job/job_controller.go

3. pkg/controller/endpointslice/endpointslice_controller.go

to name a few. I think it has substantial coverage to cause some performance improvement by using readLock(). I can go ahead and do some cpu profiling. Let me know your thoughts.
